data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/PeerForwarder.java

```diff
@@ -85,72 +93,74 @@ public PeerForwarder(final PluginSetting pluginSetting) {
         this(
                 pluginSetting,
+                new OTelProtoCodec.OTelProtoEncoder(),
                 peerForwarderConfig.getPeerClientPool(),
                 peerForwarderConfig.getHashRing(),
                 peerForwarderConfig.getMaxNumSpansPerRequest()
         );
     }
 
     @Override
-    public List<Record<ExportTraceServiceRequest>> doExecute(final Collection<Record<ExportTraceServiceRequest>> records) {
-        final Map<String, List<ResourceSpans>> groupedRS = new HashMap<>();
-
+    public List<Record<Span>> doExecute(final Collection<Record<Span>> records) {
+        final Collection<Span> spans = records.stream().map(Record::getData).collect(Collectors.toList());
+        final Map<String, List<Span>> spansByTraceId = PeerForwarderUtils.splitByTrace(spans);
         // Group ResourceSpans by consistent hashing of traceId
-        for (final Record<ExportTraceServiceRequest> record : records) {
-            for (final ResourceSpans rs : record.getData().getResourceSpansList()) {
-                final List<Map.Entry<String, ResourceSpans>> rsBatch = PeerForwarderUtils.splitByTrace(rs);
-                for (final Map.Entry<String, ResourceSpans> entry : rsBatch) {
-                    final String traceId = entry.getKey();
-                    final ResourceSpans newRS = entry.getValue();
-                    final String dataPrepperIp = hashRing.getServerIp(traceId).orElse(StaticPeerListProvider.LOCAL_ENDPOINT);
-                    groupedRS.computeIfAbsent(dataPrepperIp, x -> new ArrayList<>()).add(newRS);
-                }
-            }
+        final Map<String, List<Span>> spansByEndPoint = new HashMap<>();
+        for (final Map.Entry<String, List<Span>> entry: spansByTraceId.entrySet()) {
+            final String traceId = entry.getKey();
+            final String dataPrepperIp = hashRing.getServerIp(traceId).orElse(StaticPeerListProvider.LOCAL_ENDPOINT);
+            spansByEndPoint.computeIfAbsent(dataPrepperIp, x -> new ArrayList<>()).addAll(entry.getValue());
         }
 
-        final List<Record<ExportTraceServiceRequest>> recordsToProcessLocally = new ArrayList<>();
-        final List<CompletableFuture<Record>> forwardedRequestFutures = new ArrayList<>();
+        final List<Record<Span>> recordsToProcessLocally = new ArrayList<>();
+        final Map<CompletableFuture<ExportTraceServiceRequest>, List<Span>> forwardedRequestFuturesToSpans = new HashMap<>();
 
-        for (final Map.Entry<String, List<ResourceSpans>> entry : groupedRS.entrySet()) {
+        for (final Map.Entry<String, List<Span>> entry : spansByEndPoint.entrySet()) {
             final TraceServiceGrpc.TraceServiceBlockingStub client = getClient(entry.getKey());
+            if (isLocalClient(client)) {
+                recordsToProcessLocally.addAll(entry.getValue().stream().map(Record::new).collect(Collectors.toList()));
```

Comment:
nit: consider breaking down high complexity statements like this into multiple statements or lines for readability.